ambry-store/src/main/java/com.github.ambry.store/IndexValue.java

```diff
@@ -13,114 +13,323 @@
+    this(size, offset, FLAGS_DEFAULT_VALUE, expiresAtMs, offset.getOffset(), operationTimeInMs, serviceId, containerId);
   }
 
-  private IndexValue(long size, Offset offset, byte flags, long expiresAtMs, long originalMessageOffset) {
+  /**
+   * Constructs IndexValue based on the args passed
+   * @param size the size of the blob that this index value refers to
+   * @param offset the {@link Offset} in the {@link Log} where the blob that this index value refers to resides
+   * @param flags the {@link Flags} that needs to be set for the Index Value
+   * @param expiresAtMs the expiration time in ms at which the blob expires
+   * @param operationTimeInMs operation time in ms of the entry
+   * @param serviceId the serviceId that this blob belongs to
+   * @param containerId the containerId that this blob belongs to
+   */
+  IndexValue(long size, Offset offset, byte flags, long expiresAtMs, long operationTimeInMs, short serviceId,
+      short containerId) {
+    this(size, offset, flags, expiresAtMs, offset.getOffset(), operationTimeInMs, serviceId, containerId);
+  }
+
+  /**
+   * Constructs IndexValue based on the args passed
+   * @param size the size of the blob that this index value refers to
+   * @param offset the {@link Offset} in the {@link Log} where the blob that this index value refers to resides
+   * @param flags the flags that needs to be set for the Index Value
+   * @param expiresAtMs the expiration time in ms at which the blob expires
+   * @param originalMessageOffset the original message offset where the Put record pertaining to a delete record exists
+   *                              in the same log segment. Set to -1 otherwise.
+   * @param operationTimeInMs operation time in ms of the entry in secs
+   * @param serviceId the serviceId that this blob belongs to
+   * @param containerId the containerId that this blob belongs to
+   */
+  private IndexValue(long size, Offset offset, byte flags, long expiresAtMs, long originalMessageOffset,
+      long operationTimeInMs, short serviceId, short containerId) {
+    this.size = size;
     this.offset = offset;
-    value = ByteBuffer.allocate(Index_Value_Size_In_Bytes);
-    value.putLong(size);
-    value.putLong(offset.getOffset());
-    value.put(flags);
-    value.putLong(expiresAtMs);
-    value.putLong(originalMessageOffset);
-    value.position(0);
+    this.flags = flags;
+    int expiryInSecs = (int) (expiresAtMs / Time.MsPerSec);
+    this.expiresAtMs = expiresAtMs != Utils.Infinite_Time ? expiryInSecs * Time.MsPerSec : Utils.Infinite_Time;
+    this.originalMessageOffset = originalMessageOffset;
+    int operationTimeInSecs = (int) (operationTimeInMs / Time.MsPerSec);
+    this.operationTimeInMs =
+        operationTimeInMs != Utils.Infinite_Time ? operationTimeInSecs * Time.MsPerSec : Utils.Infinite_Time;
+    this.serviceId = serviceId;
+    this.containerId = containerId;
+    version = PersistentIndex.VERSION_1;
```

Comment:
maintain a `currentVersion` variable. That way, if we need to change it in the future, we can change it in one place.